# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/VideoSegment.cpp
# Commit: ba640de8df47
# Full Hash: ba640de8df47d54906c708eb78f2d000d6f6d7f6
# Author: Andrew Osmond <aosmond@mozilla.com>
# Date: 2024-09-13 21:45:07
# Regressor Bug: 1908245
# File Overlap Count: 1
# Description:
#   Bug 1908245 - Add support for missing requestVideoFrameCallback parameters. r=media-playback-reviewers,webidl,padenot,saschanaz
#   
#   This patch adds support for exposing the capture time, receive time,
#   processing duration and RTP timestamp parameters on the callback for
#   rVFC for WebRTC related elements.
# ==============================================================================

diff -r 2f4f9cce6f7d -r ba640de8df47 dom/media/VideoSegment.cpp
--- a/dom/media/VideoSegment.cpp	Fri Sep 13 18:15:18 2024 +0000
+++ b/dom/media/VideoSegment.cpp	Fri Sep 13 18:19:57 2024 +0000
@@ -84,12 +84,74 @@
   return image.forget();
 }
 
+void VideoSegment::AppendFrame(const VideoChunk& aChunk,
+                               const Maybe<bool>& aForceBlack,
+                               const Maybe<TimeStamp>& aTimeStamp) {
+  VideoChunk* chunk = AppendChunk(0);
+  chunk->mTimeStamp = aTimeStamp ? *aTimeStamp : aChunk.mTimeStamp;
+  chunk->mProcessingDuration = aChunk.mProcessingDuration;
+  chunk->mMediaTime = aChunk.mMediaTime;
+  chunk->mWebrtcCaptureTime = aChunk.mWebrtcCaptureTime;
+  chunk->mWebrtcReceiveTime = aChunk.mWebrtcReceiveTime;
+  chunk->mRtpTimestamp = aChunk.mRtpTimestamp;
+  VideoFrame frame(do_AddRef(aChunk.mFrame.GetImage()),
+                   aChunk.mFrame.GetIntrinsicSize());
+  MOZ_ASSERT_IF(!IsNull(), !aChunk.mTimeStamp.IsNull());
+  frame.SetForceBlack(aForceBlack ? *aForceBlack
+                                  : aChunk.mFrame.GetForceBlack());
+  frame.SetPrincipalHandle(aChunk.mFrame.GetPrincipalHandle());
+  chunk->mFrame.TakeFrom(&frame);
+}
+
 void VideoSegment::AppendFrame(already_AddRefed<Image>&& aImage,
                                const IntSize& aIntrinsicSize,
                                const PrincipalHandle& aPrincipalHandle,
-                               bool aForceBlack, TimeStamp aTimeStamp) {
+                               bool aForceBlack, TimeStamp aTimeStamp,
+                               media::TimeUnit aProcessingDuration,
+                               media::TimeUnit aMediaTime) {
+  VideoChunk* chunk = AppendChunk(0);
+  chunk->mTimeStamp = aTimeStamp;
+  chunk->mProcessingDuration = aProcessingDuration;
+  chunk->mMediaTime = aMediaTime;
+  VideoFrame frame(std::move(aImage), aIntrinsicSize);
+  MOZ_ASSERT_IF(!IsNull(), !aTimeStamp.IsNull());
+  frame.SetForceBlack(aForceBlack);
+  frame.SetPrincipalHandle(aPrincipalHandle);
+  chunk->mFrame.TakeFrom(&frame);
+}
+
+void VideoSegment::AppendWebrtcRemoteFrame(
+    already_AddRefed<Image>&& aImage, const IntSize& aIntrinsicSize,
+    const PrincipalHandle& aPrincipalHandle, bool aForceBlack,
+    TimeStamp aTimeStamp, media::TimeUnit aProcessingDuration,
+    uint32_t aRtpTimestamp, int64_t aWebrtcCaptureTimeNtp,
+    int64_t aWebrtcReceiveTimeUs) {
   VideoChunk* chunk = AppendChunk(0);
   chunk->mTimeStamp = aTimeStamp;
+  chunk->mProcessingDuration = aProcessingDuration;
+  if (aWebrtcCaptureTimeNtp > 0) {
+    chunk->mWebrtcCaptureTime = AsVariant(aWebrtcCaptureTimeNtp);
+  }
+  if (aWebrtcReceiveTimeUs > 0) {
+    chunk->mWebrtcReceiveTime = Some(aWebrtcReceiveTimeUs);
+  }
+  chunk->mRtpTimestamp = Some(aRtpTimestamp);
+  VideoFrame frame(std::move(aImage), aIntrinsicSize);
+  MOZ_ASSERT_IF(!IsNull(), !aTimeStamp.IsNull());
+  frame.SetForceBlack(aForceBlack);
+  frame.SetPrincipalHandle(aPrincipalHandle);
+  chunk->mFrame.TakeFrom(&frame);
+}
+
+void VideoSegment::AppendWebrtcLocalFrame(
+    already_AddRefed<Image>&& aImage, const IntSize& aIntrinsicSize,
+    const PrincipalHandle& aPrincipalHandle, bool aForceBlack,
+    TimeStamp aTimeStamp, media::TimeUnit aProcessingDuration,
+    TimeStamp aWebrtcCaptureTime) {
+  VideoChunk* chunk = AppendChunk(0);
+  chunk->mTimeStamp = aTimeStamp;
+  chunk->mProcessingDuration = aProcessingDuration;
+  chunk->mWebrtcCaptureTime = AsVariant(aWebrtcCaptureTime);
   VideoFrame frame(std::move(aImage), aIntrinsicSize);
   MOZ_ASSERT_IF(!IsNull(), !aTimeStamp.IsNull());
   frame.SetForceBlack(aForceBlack);